core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java

```diff
@@ -596,17 +600,37 @@ public long getKeyPrefix() {
    *
    * TODO: support forced spilling
    */
-  public UnsafeSorterIterator getIterator() throws IOException {
+  public UnsafeSorterIterator getIterator(int startIndex) throws IOException {
     if (spillWriters.isEmpty()) {
       assert(inMemSorter != null);
-      return inMemSorter.getSortedIterator();
+      UnsafeSorterIterator iter = inMemSorter.getSortedIterator();
+      for (int i = 0; i < startIndex && iter.hasNext(); i++) {
+        iter.loadNext();
+      }
+      return iter;
     } else {
       LinkedList<UnsafeSorterIterator> queue = new LinkedList<>();
+      int i = 0;
       for (UnsafeSorterSpillWriter spillWriter : spillWriters) {
-        queue.add(spillWriter.getReader(serializerManager));
+        if (i + spillWriter.recordsSpilled() <= startIndex) {
+          i += spillWriter.recordsSpilled();
+          continue;
+        } else {
+          UnsafeSorterIterator iter = spillWriter.getReader(serializerManager);
+          for (int j = i; j < startIndex && iter.hasNext(); j++) {
+            iter.loadNext();
+          }
+          queue.add(iter);
+          i += spillWriter.recordsSpilled();
+        }
       }
       if (inMemSorter != null) {
-        queue.add(inMemSorter.getSortedIterator());
+        UnsafeSorterIterator iter = inMemSorter.getSortedIterator();
```

Comment:
This is almost similar to the code in the first if clause. Let's try to get rid of this duplication, and move this logic into a separate function.